### PR TITLE
DPDK: DPDK backend does not accepts token length of more than 63 chars

### PR DESCRIPTION
**Pull request by kamleshbhalui**
so make sure token length is always less or equal to 63 chars, including dots('.') in member expression.

Currently we allow only 30 chars ,  so headers member expression like below will be within 63.
h(1).name0(30).name1(30) => total length become 63 including dots.

### CODE REVIEW

**Comment by usha1830** [comment]
@kamleshbhalui Instead of shortening, can we not use some mapping of names? After shortening the names look incomplete and also having the number fixed to 15 or any other number doesn't look like an elegant solution.

My suggestion is to have an internal mapping of the names 
<original big Name, fixed name with unique_id suffix>
<metadataxxxxxxxxxxxxxxxxxxxxxxxxxbigname, dpdk_meta1>
<metadataxxxxxxxxxxxxxxxxxxxxxxxxxbigname2, dpdk_meta2>

And as discussed,  we can emit the original name as a comment.

**Review by usha1830 — CHANGES_REQUESTED** [review]
Please see my comment for alternative approach.

**Review by psivanup — APPROVED** [review]
Functionally this fix looks good to me. As mentioned by @usha1830 , you can use different naming.

File: backends/dpdk/dpdkAsmOpt.h [hunk 1]
```diff
@@ -159,6 +159,100 @@ class RemoveUnusedMetadataFields : public Transform {
...
 
+// This pass shorten the Identifier length
+class ShortenTokenLength : public Transform {
+    ordered_map<cstring, cstring> newNameMap;
+    ordered_map<cstring, cstring> &origNameMap;
+    static size_t count;
+    // Currently Dpdk allows Identifier of 63 char long or less
+    // including dots(.) for member exp.
+    // worst case member expression will look like
+    // 1.30.30 => 63(including dot(.))
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.h:170]: This looks too conservative for metadata struct fields where the access is of the form "m.<fieldname>"
Many of the metadata fields (most of the std PSA/PNA metadata ) is being changed because of this logic. This is something not desirable.
Maybe we should handle metadata and header struct fields spearately.

File: backends/dpdk/dpdkAsmOpt.h [hunk 2]
```diff
@@ -159,6 +159,100 @@ class RemoveUnusedMetadataFields : public Transform {
...
+        newNameMap.insert(std::pair<cstring, cstring>(str, newStr));
+        origNameMap.insert(std::pair<cstring, cstring>(newStr, str));
+        return newStr;
+    }
+
+ public:
+    ShortenTokenLength(ordered_map<cstring, cstring> &origNameMap)
+        : origNameMap{origNameMap} {}
+    const IR::Node* preorder(IR::Member *m) override {
+        m->member = shortenString(m->member);
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.h:191]: We should consider the total length including m->expr.

File: backends/dpdk/dpdk.def [hunk 3]
```diff
@@ -43,20 +43,20 @@ class DpdkTable {
...
+    std::ostream& toSpec(std::ostream& out, const ordered_map<cstring, cstring>& origNameMap) const;
 #nodbprint
 #novalidate
 }
 
 class DpdkSelector {
     inline cstring name;
-    inline cstring group_id;
-    inline cstring member_id;
+    Expression group_id;
```
Comment (usha1830) [inline backends/dpdk/dpdk.def:53]: Why is this changed?

File: testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec [hunk 4]
```diff
@@ -28,8 +28,10 @@ struct next_hop_arg_t {
 }
 
 struct main_metadata_t {
-	bit<32> pna_main_input_metadata_input_port
-	bit<32> pna_main_output_metadata_output_port
+	;oldname:pna_main_input_metadata_input_port
+	bit<32> pna_main_input_metadata_input0
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec:32]: This shouldn't change and likewise many other spec files

**Review by usha1830 — CHANGES_REQUESTED** [review]
Some lint issues need to be fixed .

File: backends/dpdk/dpdkAsmOpt.h (hunk 2 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.h:191]: suppose total length is more than 63 , and it;s in form h.member1.member2, then which member will be changed member1 or member2?

File: backends/dpdk/dpdk.def (hunk 3 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdk.def:53]: otherwise visitor will not pickup this.

File: testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec (hunk 4 above)
Comment (kamleshbhalui) [reply testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec:32]: currently we allowing only 30 char, so it's changing.

File: backends/dpdk/dpdkAsmOpt.h (hunk 2 above)
Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.h:191]: member2 only should change but it should change only when the total length exceeds the length

Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.h:191]: what if member1 itself is longer than allowed(60), changing member2 won't help.

File: backends/dpdk/dpdkAsmOpt.h (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.h:170]: tried doing same

File: backends/dpdk/dpdkAsmOpt.h [hunk 5]
```diff
@@ -159,6 +159,121 @@ class RemoveUnusedMetadataFields : public Transform {
...
+    const IR::Node* preorder(IR::DpdkStructType *s) override {
+        for (auto annot : s->annotations->annotations)
+            if (annot->name == "__packet_data__") {
+                s->name = shortenString(s->name);
+                for (auto field : s->fields) {
+                    IR::StructField *f = const_cast<IR::StructField *>(field);
+                    f->name = shortenString(f->name, 30);
+                    if (auto st = f->to<IR::DpdkStructType>()) {
+                        for (auto field1 : st->fields) {
+                             IR::StructField *f1 = const_cast<IR::StructField *>(field1);
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.h:207]: const_cast is discouraged 
https://github.com/p4lang/p4c/blob/main/docs/README.md#coding-conventions

You can try creating a copy and updating it.

File: backends/dpdk/dpdkAsmOpt.h [hunk 6]
```diff
@@ -159,6 +159,121 @@ class RemoveUnusedMetadataFields : public Transform {
...
+    }
+
+    const IR::Node* preorder(IR::DpdkStructType *s) override {
+        for (auto annot : s->annotations->annotations)
+            if (annot->name == "__packet_data__") {
+                s->name = shortenString(s->name);
+                for (auto field : s->fields) {
+                    IR::StructField *f = const_cast<IR::StructField *>(field);
+                    f->name = shortenString(f->name, 30);
+                    if (auto st = f->to<IR::DpdkStructType>()) {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.h:205]: Please add a comment to describe your intent here.

**Review by usha1830 — APPROVED** [review]
I am approving this. Please take care of few minor comments before merging
